nixos/modules/services/web-apps/vikunja.nix

```diff
@@ -0,0 +1,235 @@
+          description = "SMTP host port.";
+        };
+        username = mkOption {
+          type = types.str;
+          default = "";
+          description = "SMTP username.";
+        };
+        password = mkOption {
+          type = types.str;
+          default = "";
+          description = ''
+            SMTP User password <option>config.database.user</option>.
+            Warning: this is stored in cleartext in the Nix store!
+            Use an environment file with <option>environmentFiles</option> instead.
+          '';
+        };
+        fromemail = mkOption {
+          type = types.str;
+          default = "";
+          description = "The default from address when sending emails.";
+        };
+        forcessl = mkOption {
+          type = types.bool;
+          default = false;
+          description = "By default, vikunja will try to connect with starttls, use this option to force it to use ssl.";
+        };
+      };
+    };
+  };
+  config = lib.mkIf cfg.enable {
+
+    systemd.services.vikunja-api = {
+      description = "vikunja-api";
+      after = [ "network.target" ] ++ lib.optional usePostgresql "postgresql.service" ++ lib.optional useMysql "mysql.service";
+      wantedBy = [ "multi-user.target" ];
+      path = [ cfg.package-api ];
+
+
+      serviceConfig = {
+        Type = "simple";
+        User = cfg.user;
+        Group = "vikunja";
+        WorkingDirectory = cfg.stateDir;
+        ExecStart = "${cfg.package-api}/bin/api";
+        Restart = "always";
+        EnvironmentFile = cfg.environmentFiles;
+      };
+    };
+
+    services.nginx.virtualHosts."${cfg.frontendHostname}" = {
+      locations = {
+        "/" = {
+          root = cfg.package-frontend;
+          tryFiles = "try_files $uri $uri/ /";
+        };
+        "~* ^/(api|dav|\.well-known)/" = {
+          proxyPass = "http://localhost:3456";
+          extraConfig = ''
+            client_max_body_size 20M;
+          '';
+        };
+      };
+    };
+
+    users.users = mkIf (cfg.user == "vikunja") {
+      vikunja = {
+        description = "Vikunja Service";
+        home = cfg.stateDir;
+        createHome = true;
+        useDefaultShell = true;
+        group = "vikunja";
+        isSystemUser = true;
+      };
+    };
+
+    users.groups.vikunja = {};
+
+    environment.etc."vikunja/config.json".source = configFile;
+
+    services.postgresql = optionalAttrs (usePostgresql && cfg.config.database.createDatabase) {
+      enable = mkDefault true;
+
+      ensureDatabases = [ cfg.config.database.database ];
+      ensureUsers = [
+        { name = cfg.config.database.user;
+          ensurePermissions = { "DATABASE ${cfg.config.database.database}" = "ALL PRIVILEGES"; };
+        }
+      ];
+    };
+
+    services.mysql = optionalAttrs (useMysql && cfg.config.database.createDatabase) {
+      enable = mkDefault true;
+      package = mkDefault pkgs.mariadb;
+
+      ensureDatabases = [ cfg.config.database.database ];
```

Comment:
I have a problem with that. I don't have much experience with mariadb. I could try to build something that sets the database charsets and collations correctly, but I won't be sure it's correct. Only postgresql I am sure about. What do you think: should I try or should I leave the creation of the database (maybe only when the user chooses mariadb) to the user? (even if this would not be a complete nixos module).

Reply:
It a basic test works and you don't have any reason to believe the encoding is wrong you could leave it as is. Maybe that is easiest for now.

Reply:
For me it would also be okay to let the user create the database

Reply:
That sounds good :+1: